Update metrics summary API

timstclair opened the pull request:
- Change resource Quantity types to *uint64
- Change CoreSeconds -> CoreNanoSeconds
- Rename NonLocalObjectReference to PodReference
- Add UID to PodReference
- Add UserDefinedMetrics

Forked from https://github.com/kubernetes/kubernetes/pull/19932
For https://github.com/kubernetes/kubernetes/issues/12483

/cc @vishh @pwittrock

vishh commented:
LGTM

timstclair commented:
Update:
Changed `CPUStats.UsageCores` to `CPUStats.UsageNanoCores`
Changed `json:"usageCoreSeconds"` to `json:"usageCoreNanoSeconds"`

timstclair commented:
Update: Changed `PageFaults` and `MajorPageFaults` type from `*int64` to `*uint64` for consistency.

smarterclayton commented:
uint64 change is unlikely to have the effect we want since most JSON clients have troubles above 52bits.    We don't use uint64 anywhere else in the API because of this - is there value in making the change here?

smarterclayton commented:
API conventions technically calls for us to use strings if we think this is important:

> All numbers (e.g., uint32, int64) are converted to float64 by Javascript and some other languages, so any field which is expected to exceed that either in magnitude or in precision (specifically integer values > 53 bits) should be serialized and accepted as strings.

timstclair commented:
@smarterclayton - There were too motivations to this change from the original resource.Quantity version. We eventually want to move most of this into a separate standalone cAdvisor pod, so we were choosing to maintain consistency with cAdvisor rather than Kubernetes. The other motivation is that there have been some performance issues around resource.Quantity in the past, and since this API will be exporting a lot of data it will be very sensitive to performance issues.

vishh commented on pkg/kubelet/server/stats/types.go:
```diff
@@ -70,12 +71,18 @@ type PodStats struct {
 	// Reference to the measured container.
 	Name string `json:"name"`
+	// The time at which data collection for this container was (re)started.
+	StartTime unversioned.Time `json:"startTime"`
```
@timstclair: Is `unversioned.Time` a portable time definition or should we stick to nanoseconds since epoch or something similar?

timstclair replied:
I think it depends on whether we want to optimize for human readability or portability. Personally I prefer readability, as I think the RFC 3339 spec that time uses is portable enough.

vishh commented on pkg/kubelet/server/stats/types.go:
```diff
@@ -139,11 +147,45 @@ type VolumeStats struct {
+	UserDefinedMetricDescriptor `json:",inline"`
+	// Value of the metric. Float64s have 53 bit precision.
+	// We do not forsee any metrics exceeding that value.
+	Value float64 `json:"value"`
```
FYI: @timstclair: This field is not optional. Just highlighting it in case you missed it.

timstclair replied:
ACK

timstclair commented:
Added UserDefinedMetrics changes we discussed offline. PTAL.

vishh commented:
LGTM

timstclair commented:
Squashed.

smarterclayton commented:
Ok, that's fine by me.  Just wanted to hear it qualified.

On Fri, Jan 29, 2016 at 2:19 PM, Tim St. Clair <notifications@github.com>
wrote:

> @smarterclayton <https://github.com/smarterclayton> - There were too
> motivations to this change from the original resource.Quantity version. We
> eventually want to move most of this into a separate standalone cAdvisor
> pod, so we were choosing to maintain consistency with cAdvisor rather than
> Kubernetes. The other motivation is that there have been some performance
> issues around resource.Quantity in the past, and since this API will be
> exporting a lot of data it will be very sensitive to performance issues.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/20305#issuecomment-176919848>
> .
>

tallclair committed:
Update metrics summary API

- Change resource Quantity types to *uint64
- Change CoreSeconds -> CoreNanoSeconds
- Rename NonLocalObjectReference to PodReference
- Add UID to PodReference
- Add UserDefinedMetrics
- Add StartTime to container and pod metrics.

timstclair commented:
Added `StartTime` to `NodeStats`
